### PR TITLE
Specify an s3uri as {bucket}/{key}?versionId={versionId}

### PR DESCRIPTION
**Pull request by mariusvniekerk**
This should in principal allow us to do things like

```
dask.dataframe.read_parquet([
    's3://bucket/somepath/file_part0001.parquet?versionId=ksfgsdkfkjadsf',
    's3://bucket/somepath/file_part0002.parquet?versionId=lijserslifhsdiu',
])
```

This should work for most things that make sense.

Not everything has proper test cases yet.

### CODE REVIEW

File: s3fs/tests/test_s3fs.py [hunk 1]
```diff
@@ -1096,6 +1098,27 @@ def test_list_versions_many(s3):
     assert len(versions) == 1200
 
 
+def test_versioned_file_fullpath(s3):
+    versioned_file = versioned_bucket_name + '/versioned_file_fullpath'
+    s3 = S3FileSystem(anon=False, version_aware=True)
```
Comment (martindurant) [inline s3fs/tests/test_s3fs.py]: Isn't this per-bucket?

Comment (mariusvniekerk) [reply s3fs/tests/test_s3fs.py]: This is a per bucket setting, but we are not detecting anywhere right now in s3fs whether a particular bucket has versioning enabled,  consequently it is just mostly used blanketly (similar in the other versioned tests)

**Comment by mariusvniekerk** [comment]
> * there is a general need across all filesystems to be able to specify more than just a URL to identify a bunch of bytes. In principle, I am OK with including a query parameter like this - as far as I know, only actual HTTP uses this for anything at the moment. It would be better if it was a known syntax elsewhere that we could refer to, though.

Let me know where in fsspec we can define with this extended url format could look like.

Reasonably okay with changing the split_path method back, but....

```
bucket, key = split_path(path)
```

is no longer guaranteed to actually return a `key` or a valid prefix.


Should these splitting methods be defined more explicitly in `fsspec`?

> * we don't actually test the path that this was set to solve, `fsspec.open_files` with versioning. Do the query fragments survive passing through `fsspec.utils.infer_storage_options`?

Yes, That is still a bit of a TODO.

> * how will this interact with other invokations of `open_files`, e.g., with `mode='wb'`, or paths with globs? Do we need to explicitly raise?

`w` of any kind should raise, since you cannot control how s3 assigns version ids.   
`globs` are largely meaningless with versions since they apply to only a single key.  There exists no fast way to get a file within a prefix with a given versionId.

**Comment by mariusvniekerk** [comment]
@martindurant green now, ready for rereview

**Comment by mariusvniekerk** [comment]
Whoops the last one we me accidentally posting as another user.

**Comment by mariusvniekerk** [comment]
@martindurant This is ready for review again

**Comment by martindurant** [comment]
Is `info(..., version_id=)` now deprecated, then?

**Comment by martindurant** [comment]
ping @mariusvniekerk

**Comment by mariusvniekerk** [comment]
@martindurant This right now still uses that method to get things like length and etag.  If a uri AND a version_id are given and they mismatch we should probably raise.

**Comment by martindurant** [comment]
> we should probably raise.

Please!

**Comment by mariusvniekerk** [comment]
@martindurant Added this functionality

**Comment by martindurant** [comment]
ooops...
```
verssion_ids.remove(None)
```
(extra s)

**Comment by martindurant** [comment]
`if len(version_ids > 1):`
->
`if len(version_ids) > 1:`

**Comment by mariusvniekerk** [comment]
> `if len(version_ids > 1):`
> ->
> `if len(version_ids) > 1:`

The amount of times i've written that....

**Comment by martindurant** [comment]
Do remind me to release this sometime.

**Pull request merged**